Mocha tests fail when browser window is not foreground window

### Check for duplicates

- [x] I have searched for similar issues before opening a new one.

### Description

When running the mocha tests, either manually or via `npm test`, many tests will fail if the browser window is not the foreground window.  This makes it difficult to (usefully) run tests in the background while doing other work, which is annoying because the Mocha tests are fairly slow.

It's not clear how this might be remedied, since presumably at least some of the focus tests require verifying actual focus, but it would be nice if e.g. the browser window not having focus would result in exactly one test failure, so that any other non-focus-related test failures will be obvious.

(This problem did not exist prior to the introduction of support for focus in Blockly, and caused me some considerable confusion as I attempted to figure out how some completely unrelated changes I was working on could possibly be breaking so many tests, and why sometimes all the tests passed fine but other times they would not.)

### Reproduction steps

1. Run `npm test`
2. When a Chrome window opens for the Mocha tests, click on another window.
3. Observe (high but non-1 probability) that numerous (typically around 95) focus-related tests fail.


### Stack trace

```shell
// Example failure:
FocusManager
  focusNode()
    fires focusin event‣
      AssertionError: expected +0 to equal 1
        at Context.<anonymous> (focus_manager_test.js:398:14)
```

### Screenshots

_No response_

### Browsers

Chrome desktop

--- a/gulpfile.mjs
+++ b/gulpfile.mjs
@@ -45,7 +45,11 @@ import {
   publishBeta,
   recompile,
 } from './scripts/gulpfiles/release_tasks.mjs';
-import {generators, test} from './scripts/gulpfiles/test_tasks.mjs';
+import {
+  generators,
+  interactiveMocha,
+  test,
+} from './scripts/gulpfiles/test_tasks.mjs';
 
 const clean = parallel(cleanBuildDir, cleanReleaseDir);
 
@@ -80,6 +84,7 @@ export {
   clean,
   test,
   generators as testGenerators,
+  interactiveMocha,
   buildAdvancedCompilationTest,
   createRC as gitCreateRC,
   docs,

--- a/package.json
+++ b/package.json
@@ -45,7 +45,7 @@
     "test": "gulp test",
     "test:browser": "cd tests/browser && npx mocha",
     "test:generators": "gulp testGenerators",
-    "test:mocha:interactive": "npm run build && concurrently -n tsc,server \"tsc --watch --preserveWatchOutput --outDir \"build/src\" --declarationDir \"build/declarations\"\" \"http-server ./ -o /tests/mocha/index.html -c-1\"",
+    "test:mocha:interactive": "npm run build && concurrently -n tsc,server \"tsc --watch --preserveWatchOutput --outDir \"build/src\" --declarationDir \"build/declarations\"\" \"gulp interactiveMocha\"",
     "test:compile:advanced": "gulp buildAdvancedCompilationTest --debug",
     "updateGithubPages": "npm ci && gulp gitUpdateGithubPages"
   },

--- a/package.json
+++ b/package.json
@@ -138,6 +138,7 @@
     "patch-package": "^8.0.0",
     "prettier": "^3.3.3",
     "prettier-plugin-organize-imports": "^4.0.0",
+    "puppeteer-core": "^24.17.0",
     "readline-sync": "^1.4.10",
     "rimraf": "^5.0.0",
     "typescript": "^5.3.3",

--- a/scripts/gulpfiles/test_tasks.mjs
+++ b/scripts/gulpfiles/test_tasks.mjs
@@ -257,9 +257,9 @@ async function metadata() {
  * Run Mocha tests inside a browser.
  * @return {Promise} Asynchronous result.
  */
-async function mocha() {
+async function mocha(exitOnCompletion = true) {
   return runTestTask('mocha', async () => {
-    const result = await runMochaTestsInBrowser().catch(e => {
+    const result = await runMochaTestsInBrowser(exitOnCompletion).catch(e => {
       throw e;
     });
     if (result) {
@@ -267,6 +267,14 @@ async function mocha() {
     }
     console.log('Mocha tests passed');
   });
+}
+
+/**
+ * Run Mocha tests inside a browser and keep the browser open upon completion.
+ * @return {Promise} Asynchronous result.
+ */
+export async function interactiveMocha() {
+  return mocha(false);
 }
 
 /**

--- a/tests/mocha/webdriver.js
+++ b/tests/mocha/webdriver.js
@@ -45,6 +45,17 @@ async function runMochaTestsInBrowser() {
   console.log('Loading URL: ' + url);
   await browser.url(url);
 
+  // Toggle the devtools setting to emulate focus, so that the window will
+  // always act as if it has focus regardless of the state of the window manager
+  // or operating system. This improves the reliability of FocusManager-related
+  // tests.
+  const puppeteer = await browser.getPuppeteer();
+  await browser.call(async () => {
+    const page = (await puppeteer.pages())[0];
+    const session = await page.createCDPSession();
+    await session.send('Emulation.setFocusEmulationEnabled', { enabled: true });
+  });
+
   await browser.waitUntil(async() => {
     const elem = await browser.$('#failureCount');
     const text = await elem.getAttribute('tests_failed');

--- a/tests/mocha/webdriver.js
+++ b/tests/mocha/webdriver.js
@@ -15,9 +15,12 @@ const {posixPath} = require('../../scripts/helpers');
  * Runs the Mocha tests in this directory in Chrome. It uses webdriverio to
  * launch Chrome and load index.html. Outputs a summary of the test results
  * to the console.
+ *
+ * @param {boolean} exitOnCompletetion True if the browser should automatically
+ *     quit after tests have finished running.
  * @return {number} 0 on success, 1 on failure.
  */
-async function runMochaTestsInBrowser() {
+async function runMochaTestsInBrowser(exitOnCompletion = true) {
   const options = {
     capabilities: {
       browserName: 'chrome',
@@ -85,7 +88,7 @@ async function runMochaTestsInBrowser() {
   if (parseInt(numOfFailure) !== 0) {
     return 1;
   }
-  await browser.deleteSession();
+  if (exitOnCompletion) await browser.deleteSession();
   return 0;
 }
 

